### ISSUE TITLE
Inspect dependencies to detect unhandled exceptions

### ISSUE THREAD

**Issue by oliversheridanmethven** [jsh9/pydoclint#281]
With `py.typed` files tools like `mypy` can look into the dependency code and check that you are using third party libraries correctly. 

The input types and return types are covered by this, but there is a definite blind spot, for example, what is raised. I imagine a world where there is a third party library called `third_party` which implements some `foo`:

```python
def foo() -> None:
    """
    Does something.

    Raises
    ------
    TypeError
        ...
    ValueError
        ...
    RuntimeError
        ...
    """
    ...
```

and then in an application code where you write some `try-catch` loop you might get

```python
from third_party import foo

# Example 1
try:
    foo()   
except TypeError:
    pass
except RuntimeError:
    pass
# Did we mean to catch ValueError here?    
    

# Example 2
try:
    foo()   
except TypeError:
    pass
except RuntimeError:
    pass
except Exception as e:  # Probably not intended to catch value-error.
    raise CustomError("An error occurred") from e
```

I suspect we could then enable options where we could handle these scenarios in the following ways:

```python
# Should raise an issue.
try:
    foo()   
except TypeError:
    pass
except RuntimeError:
    pass

# Should not raise any issue.
try: # pydoclint: ignore[missing-exception=ValueError]
    foo()   
except TypeError:
    pass
except RuntimeError:
    pass
    

# Should raise an issue.
try:
    foo()   
except TypeError:
    pass
except RuntimeError:
    pass
except Exception as e: 
    raise CustomError("An error occurred") from e

# Should not raise an issue.
try: # pydoclint: ignore[missing-exception=ValueError]
    foo()   
except TypeError:
    pass
except RuntimeError:
    pass
except Exception as e:  
    raise CustomError("An error occurred") from e

# Should not raise an issue if pydoclint is passed the option --warn_catch_all_exceptions=false
try: # pydoclint: ignore[missing-exception=ValueError]
    foo()   
except TypeError:
    pass
except RuntimeError:
    pass
except Exception as e:  
    raise CustomError("An error occurred") from e
```

To my knowledge, the idea of [allowing functions to declare what they raise in their signature is un-implemented (and strongly opposed by Guido)](https://stackoverflow.com/a/44282299/5134817), so I think the only place this information resides is perhaps in the docstring, and if we inspect our dependencies it should be possible to catch all sort of places where we are not handling behaviour declared in the documentation of the dependencies. I suspect if there is a missing docstring or one that does not have a raises section, then something akin to a `Any` would be sensible. Similarly, pydoclint should not check the dependencies are correct. 

Of course, another way to tackle this is to instead inspect the source code, in which case then a conventional static analyser like `mypy` could tackle this, although I am not sure if they plan to adopt this feature.

**Comment by jsh9** [comment]
Hi, thanks for the suggestion, but this is the kind of functionality that _pydoclint_ chooses not to implement.

According to Section 1 of [Notes for Users](https://github.com/jsh9/pydoclint/blob/main/docs/notes_for_users.md#1-why-is-pydoclint-so-much-faster-than-darglint-hhe), _pydoclint_ is very fast because it only does static code analysis rather than dynamic.  The feature you're requesting requires dynamic code analysis (building the whole projects, installing all dependencies, fetching the underlying code, ...).  This would slow down _pydoclint_ significantly.

**Comment by oliversheridanmethven** [comment]
I am not entirely sure it is dynamic analysis, as the code doesn't need to run, it seems to be it's just *more* static analysis over more files and dependencies. Static analysers like MyPy and PyRight do very similar sorts of things. It is definitely true though that to check the validity of some code required parsing files (possibly very many) from outside of that file, so it increases the workload a large margin. 

As for the speed concern, well, if it's a concern, users (like me) can always opt into the extra behaviour and bite the bullet for the sake of extra peace of mind.

**Comment by jsh9** [comment]
Speed is a huge concern, and I think implementing it is far from trivial (I may be wrong).  If you have a way to implement this feature in a clean, modular, and maintainable way, you're very welcome to submit a PR.  (But I still may not accept the PR if the user experience is not "this just works".)
